src/System.Memory/src/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -0,0 +1,377 @@
+        {
+            Debug.Assert(startSegment != null);
+            Debug.Assert(endSegment != null);
+            Debug.Assert(startSegment.Memory.Length >= startIndex);
+            Debug.Assert(endSegment.Memory.Length >= endIndex);
+
+            BufferStart = new SequencePosition(startSegment, startIndex | MemoryListStartMask);
+            BufferEnd = new SequencePosition(endSegment, endIndex | MemoryListEndMask);
+        }
+
+        public ReadOnlyBuffer(T[] array) : this(array, 0, array.Length)
+        {
+        }
+
+        public ReadOnlyBuffer(T[] array, int offset, int length)
+        {
+            if (array == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.array);
+            }
+
+            BufferStart = new SequencePosition(array, offset | ArrayStartMask);
+            BufferEnd = new SequencePosition(array, offset + length | ArrayEndMask);
+        }
+
+        // Consumer is expected to manage lifetime of memory until
+        // ReadOnlyBuffer is not used anymore
+        public ReadOnlyBuffer(Memory<T> memory)
+        {
+            var segment = new ReadOnlyBufferSegment
+            {
+                Memory = memory
+            };
+            BufferStart = new SequencePosition(segment, 0 | MemoryListStartMask);
+            BufferEnd = new SequencePosition(segment, memory.Length | MemoryListEndMask);
+        }
+
+        public ReadOnlyBuffer(IEnumerable<Memory<T>> buffers)
+        {
+            ReadOnlyBufferSegment segment = null;
+            ReadOnlyBufferSegment first = null;
+            foreach (var buffer in buffers)
+            {
+                var newSegment = new ReadOnlyBufferSegment()
+                {
+                    Memory = buffer,
+                    RunningIndex = segment?.RunningIndex ?? 0
+                };
+
+                if (segment != null)
+                {
+                    segment.Next = newSegment;
+                }
+                else
+                {
+                    first = newSegment;
+                }
+
+                segment = newSegment;
+            }
+
+            if (first == null)
+            {
+                first = segment = new ReadOnlyBufferSegment();
+            }
+
+            BufferStart = new SequencePosition(first, 0 | MemoryListStartMask);
+            BufferEnd = new SequencePosition(segment, segment.Memory.Length | MemoryListEndMask);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given <see cref="ReadOnlyBuffer{T}"/>, beginning at <paramref name="offset"/>, and is at most <paramref name="length"/> bytes
+        /// </summary>
+        /// <param name="offset">The index at which to begin this slice.</param>
+        /// <param name="length">The length of the slice</param>
+        public ReadOnlyBuffer<T> Slice(long offset, long length)
+        {
+            var begin = Seek(BufferStart, BufferEnd, offset, false);
```

Comment:
State type here (instead of var) and anywhere else where the type cannot be statically inferred by the right hand expression.